src/src/Helpers/TarStream.cpp

```diff
@@ -0,0 +1,603 @@
+
+        memcpy(&_tarData[_headerPosition], &buf[bufOffset], toMove);
+        _headerPosition += toMove;
+        bufOffset       += toMove;
+
+        if (_headerPosition == TAR_HEADER_SIZE) {
+          bool isBlank = true;
+
+          for (size_t n = 0; n < TAR_HEADER_SIZE && isBlank; n++) {
+            isBlank &= (_tarData[n] == 0u);
+          }
+
+          if (isBlank) {
+            _headerPosition = 0;
+            _streamState    = TarStreamState_e::WritingFinal;
+            # if TAR_STREAM_DEBUG
+            addLog(LOG_LEVEL_INFO, concat(F("TarStream: Switch from "), F("Initial/WritingHeader to WritingFinal")));
+            # endif // if TAR_STREAM_DEBUG
+          } else {
+            String fname(_tarHeader.name);
+            size_t fsize   = strtoul(_tarHeader.size, nullptr, 8); // Octal
+            bool   isValid = validateHeader();
+
+            # if TAR_STREAM_DEBUG
+            addLog(LOG_LEVEL_INFO,
+                   strformat(F("%sWrite Receiving file %s size: %d"), String(F("TarStream: ")).c_str(), fname.c_str(), fsize));
+            # endif // if TAR_STREAM_DEBUG
+
+            if ((_tarHeader.typeflag == REGTYPE) &&
+                isValid) {           // Checked: typeflag, magic & checksum
+              addFile(fname, fsize); // Add to list
+              _fileIndex++;
+              _filesSizes += fsize;
+              bool validConfig = true;
+
+              bufOffset += TAR_BLOCK_SIZE - TAR_HEADER_SIZE; // Skip remaining bytes to start of next block
+
+              if (matchFileType(fname, FileType::CONFIG_DAT)) {
+                validConfig = validateUploadConfigDat(&buf[bufOffset]);
+              }
+
+              if (validConfig) {
+                _streamState = TarStreamState_e::WritingFile;
+
+                size_t available = UINT32_MAX;
+
+                if (FileDestination_e::SD != _destination) { // Check flash storage only
+                  available = SpiffsFreeSpace();
+                  fs::File tmpfile = tryOpenFile(patch_fname(_filesList[_fileIndex].fileName).c_str(), "r", _destination);
+
+                  if (tmpfile) {
+                    available += tmpfile.size();                  // Existing file will be deleted
+                    tmpfile.close();
+                  }
+                  available -= 2 * SpiffsBlocksize();             // Leave at least 2 blocks free after save
+                }
+
+                if (available > _filesList[_fileIndex].tarSize) { // Use rounded-up size
+                  // delete and create file for write mode
+                  if (!tryDeleteFile(patch_fname(_filesList[_fileIndex].fileName), _destination)) {
+                    addLog(LOG_LEVEL_ERROR, concat(F("TarStream: "), concat(F("Can't delete file: "), _filesList[_fileIndex].fileName)));
+                  }
+                  _currentFile = tryOpenFile(patch_fname(_filesList[_fileIndex].fileName).c_str(), "w", _destination);
```

Comment:
`tryOpenFile` accepts a `String` as argument.
By using `.c_str()` you essentially copy the string as you're calling the `String(const char*)` constructor.  (probably also on other functions from Storage)